File: fm-dddrive/src/main/java/io/dddrive/core/property/model/PartListProperty.java
package io.dddrive.core.property.model;

import java.util.List;

import io.dddrive.core.ddd.model.Part;

public interface PartListProperty<P extends Part<?>> extends Property<P> {

	Class<P> getPartType();

	Integer getPartCount();

	P getPart(Integer seqNr);

	P getPartById(Integer partId);

	List<P> getParts();

	void clearParts();

	P addPart(Integer partId);

	void removePart(Integer partId);

	void removePart(P part);

	/**
	 * Returns the index of the specified part within this list.
	 *
	 * @param part the part to find
	 * @return the index of the part, or -1 if the part is not in this list.
	 */
	int getIndexOfPart(Part<?> part);

}

File: fm-dddrive/src/main/java/io/dddrive/property/model/PartListProperty.kt
package io.dddrive.property.model

import io.dddrive.ddd.model.Part

/**
 * Property that holds a list of parts.
 *
 * Implements [Iterable] and [Collection] interfaces so it can be used directly as a list in consumer code.
 */
interface PartListProperty<P : Part<*>> :
	Property<P>,
	Iterable<P>,
	Collection<P> {

	val partType: Class<P>

	override val size: Int

	operator fun get(seqNr: Int): P

	fun getById(partId: Int): P

	/**
	 * Removes all parts from this list.
	 * Alias for [clear].
	 */
	fun clear()

	fun add(partId: Int? = null): P

	fun remove(partId: Int)

	fun remove(part: P)

	/**
	 * Returns the index of the specified part within this list.
	 *
	 * @param part the part to find
	 * @return the index of the part, or -1 if the part is not in this list.
	 */
	fun indexOf(part: P): Int

}
